pkg/ring/ring.go

```diff
@@ -512,18 +518,24 @@ func (r *Ring) GetReplicationSetForOperation(op Operation) (ReplicationSet, erro
 	maxUnavailableZones := 0
 
 	if r.cfg.ZoneAwarenessEnabled {
-		// Given data is replicated to RF different zones, we can tolerate a number of
-		// RF/2 failing zones. However, we need to protect from the case the ring currently
-		// contains instances in a number of zones < RF.
 		numReplicatedZones := utilmath.Min(len(r.ringZones), r.cfg.ReplicationFactor)
-		minSuccessZones := (numReplicatedZones / 2) + 1
-		maxUnavailableZones = minSuccessZones - 1
+		if requireQuorum {
+			// Given data is replicated to RF different zones, we can tolerate a number of
+			// RF/2 failing zones. However, we need to protect from the case the ring currently
+			// contains instances in a number of zones < RF.
+			minSuccessZones := (numReplicatedZones / 2) + 1
+			maxUnavailableZones = minSuccessZones - 1
+		} else {
+			// Given that quorum is not required, we only need at least one of the zone to be healthy to succeed. But we
+			// also need to handle case when RF < number of zones.
+			maxUnavailableZones = numReplicatedZones - 1
+		}
 
 		if len(zoneFailures) > maxUnavailableZones {
-			return ReplicationSet{}, ErrTooManyUnhealthyInstances
+			return ReplicationSet{}, zoneFailures, ErrTooManyUnhealthyInstances
 		}
 
-		if len(zoneFailures) > 0 {
+		if requireQuorum && len(zoneFailures) > 0 {
```

Comment:
Do we think it is safe to query zones with failed instances if quorum is not required?

Reply:
For the case of ruler it will be necessary because healthy rulers in the same zone as the failed instance could be assigned to evaluate rule groups so we would want to get the rule groups from those rulers instead of the backup.